lib/IDE/CodeCompletionResultType.cpp

```diff
@@ -11,50 +11,227 @@
 //===----------------------------------------------------------------------===//
 
 #include "swift/IDE/CodeCompletionResultType.h"
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/ProtocolConformance.h"
+#include "swift/AST/USRGeneration.h"
 #include "swift/Sema/IDETypeChecking.h"
 
 using namespace swift;
 using namespace ide;
+using TypeRelation = CodeCompletionResultTypeRelation;
 
-static CodeCompletionResultTypeRelation
-calculateTypeRelation(Type Ty, Type ExpectedTy, const DeclContext &DC) {
+// MARK: - USRBasedTypeContext
+
+USRBasedTypeContext::USRBasedTypeContext(const ExpectedTypeContext *TypeContext,
+                                         const DeclContext *DC,
```

Comment:
Is `DC` used only for getting `Void` type?
If so, I'd suggest to use `Type::getASTContext()` from one of the expected types. Passing `DC` looks just confusing.